File: src/main/core/sources/thunderstore.js
'use strict';

/**
 * Каталог Thunderstore (Lethal Company).
 *
 * До 1.6 ModHub качал весь список пакетов сообщества (/api/v1/package/).
 * Для Lethal Company это 35 МБ в сжатом виде и сотни мегабайт после
 * распаковки: окно подвисало, запрос не укладывался во время ожидания,
 * и каталог оставался пустым. Теперь — API нового сайта Thunderstore
 * (cyberstorm): по 20 пакетов на страницу, сразу отсортированные, с поиском
 * на стороне сервера. Подробности пакета, зависимости и описание берутся
 * отдельными маленькими запросами, только когда человек их открыл.
 */

const { fetchJson } = require('../download');

const BASE = 'https://thunderstore.io';
const DETAIL_TTL = 15 * 60 * 1000;
const PARALLEL = 6;

const ORDERINGS = {
  popular: 'most-downloaded',
  rating: 'top-rated',
  new: 'newest',
  updated: 'last-updated',
};

const detailCache = new Map(); // "community/Namespace-Name" -> { at, mod }

/** "Namespace-Name" -> { namespace, name }. Дефис в именах Thunderstore запрещён. */
function splitId(id) {
  const text = String(id ?? '');
  const dash = text.indexOf('-');
  if (dash <= 0 || dash === text.length - 1) return null;
  return { namespace: text.slice(0, dash), name: text.slice(dash + 1) };
}

function isNotFound(error) {
  return /^404\b/.test(String(error?.message ?? ''));
}

/** Ограничитель параллельных запросов: дерево зависимостей сборки бывает на сотню пакетов. */
let active = 0;
const queue = [];
async function limited(task) {
  if (active >= PARALLEL) await new Promise((resolve) => queue.push(resolve));
  active += 1;
  try {
    return await task();
  } finally {
    active -= 1;
    queue.shift()?.();
  }
}

/** Приводит пакет Thunderstore к внутреннему виду ModHub. */
function toMod(item, community) {
  const communityId = item.community_identifier ?? community;
  return {
    source: 'thunderstore',
    id: `${item.namespace}-${item.name}`,
    name: String(item.name ?? '').replace(/_/g, ' '),
    author: item.namespace ?? '',
    version: item.latest_version_number ?? '',
    description: item.description ?? '',
    icon: item.icon_url ?? null,
    // Значки Thunderstore всегда квадратные 256×256: в карточке их ставим
    // плиткой на размытом фоне, а не растягиваем на всю ширину.
    iconShape: 'square',
    url: `${BASE}/c/${communityId}/p/${item.namespace}/${item.name}/`,
    downloads: Number(item.download_count ?? 0),
    rating: Number(item.rating_count ?? 0),
    updatedAt: item.last_updated ?? item.version_created ?? null,
    categories: (item.categories ?? []).map((c) => c?.name).filter(Boolean),
    pinned: Boolean(item.is_pinned),
    dependencies: [],
  };
}

/**
 * Страница каталога.
 * @returns {Promise<{mods: object[], total: number, hasMore: boolean, page: number}>}
 */
async function search(community, query = '', options = {}) {
  const page = Math.max(1, Number(options.page) || 1);
  const params = new URLSearchParams({ page: String(page), ordering: ORDERINGS[options.sort] ?? ORDERINGS.popular });
  const needle = String(query ?? '').trim();
  if (needle) params.set('q', needle);

  const data = await fetchJson(`${BASE}/api/cyberstorm/listing/${encodeURIComponent(community)}/?${params}`, {
    timeout: 20000,
  });
  const mods = (data?.results ?? [])
    .filter((item) => item && !item.is_nsfw && !item.is_deprecated)
    .map((item) => toMod(item, community));
  return { mods, total: Number(data?.count ?? mods.length), hasMore: Boolean(data?.next), page };
}

/** Полная карточка пакета: версия, ссылка на архив, зависимости. */
async function getById(community, id) {
  const parts = splitId(id);
  if (!parts) return null;

  const key = `${community}/${id}`;
  const cached = detailCache.get(key);
  if (cached && Date.now() - cached.at < DETAIL_TTL) return cached.mod;

  let data;
  try {
    data = await limited(() =>
      fetchJson(
        `${BASE}/api/cyberstorm/listing/${encodeURIComponent(community)}/${encodeURIComponent(parts.namespace)}/${encodeURIComponent(parts.name)}/`,
        { timeout: 20000 }
      )
    );
  } catch (error) {
    if (isNotFound(error)) return null;
    throw error;
  }

  const deps = (data?.dependencies ?? []).filter((d) => d?.namespace && d?.name);
  const version = data?.latest_version_number ?? '';
  const mod = {
    ...toMod(data, community),
    version,
    downloadUrl: data?.download_url || `${BASE}/package/download/${parts.namespace}/${parts.name}/${version}/`,
    website: data?.website_url || null,
    fileSize: Number(data?.size ?? 0) || null,
    dependencies: deps.filter((d) => !d.is_removed).map((d) => `${d.namespace}-${d.name}`),
    requirements: deps.map((d) => ({
      id: `${d.namespace}-${d.name}`,
      name: String(d.name).replace(/_/g, ' '),
      icon: d.icon_url ?? null,
      available: !d.is_removed && !d.is_unavailable,
    })),
  };
  detailCache.set(key, { at: Date.now(), mod });
  return mod;
}

/** Описание пакета — README, уже переведённый сайтом в HTML. */
async function readme(id) {
  const parts = splitId(id);
  if (!parts) return null;
  try {
    const data = await fetchJson(
      `${BASE}/api/cyberstorm/package/${encodeURIComponent(parts.namespace)}/${encodeURIComponent(parts.name)}/latest/readme/`,
      { timeout: 20000 }
    );
    return data?.html ? { format: 'html', content: data.html, base: `${BASE}/` } : null;
  } catch (error) {
    if (isNotFound(error)) return null;
    throw error;
  }
}

/**
 * Разворачивает дерево зависимостей в плоский список для установки.
 *
 * Именно этот шаг убирает главный источник сломанных сборок: автор упомянул
 * нужную библиотеку одной строкой в описании, человек её не заметил, игра
 * не запустилась. Здесь зависимости машиночитаемые, так что гадать не нужно.
 *
 * @returns {Promise<{order: object[], missing: string[]}>} зависимости раньше зависимых
 */
async function resolveDependencies(community, rootId) {
  const order = [];
  const missing = [];
  const tasks = new Map();

  function visit(id, trail) {
    if (trail.includes(id)) return Promise.resolve(); // цикл — просто не зацикливаемся
    if (tasks.has(id)) return tasks.get(id);
    const task = (async () => {
      const mod = await getById(community, id).catch(() => null);
      if (!mod) {
        missing.push(id);
        return;
      }
      await Promise.all(mod.dependencies.map((dep) => visit(dep, [...trail, id])));
      order.push(mod);
    })();
    tasks.set(id, task);
    return task;
  }

  await visit(rootId, []);
  return { order, missing };
}

async function categories() {
  return [];
}

module.exports = {
  search,
  getById,
  readme,
  categories,
  resolveDependencies,
  splitId,
  toMod,
};

File: desktop/ModLaunch/Sources/Thunderstore.cs
using System.Collections.Concurrent;
using System.Text.Json.Nodes;
using ModLaunch.Core;

namespace ModLaunch.Sources;

/// <summary>
/// Каталог Thunderstore: тот же открытый API, которым пользуется сам сайт
/// (cyberstorm) — постранично, с сортировкой и категориями.
/// </summary>
public static class Thunderstore
{
    const string Base = "https://thunderstore.io";
    static readonly ConcurrentDictionary<string, (DateTime At, ModInfo Mod)> Details = new();
    static readonly ConcurrentDictionary<string, Dictionary<string, string>> CategoryMaps = new();
    static readonly SemaphoreSlim Parallel = new(6);

    static string Ordering(SortBy sort) => sort switch
    {
        SortBy.Rating => "top-rated",
        SortBy.New => "newest",
        SortBy.Updated => "last-updated",
        _ => "most-downloaded",
    };

    public static (string Ns, string Name)? Split(string id)
    {
        // Имя пакета — без дефисов, а в имени автора они бывают (FunkFrog-and-Sipondo): делим по последнему.
        var dash = id.LastIndexOf('-');
        return dash <= 0 || dash == id.Length - 1 ? null : (id[..dash], id[(dash + 1)..]);
    }

    static ModInfo ToMod(JsonNode item, string community)
    {
        var ns = item.Str("namespace") ?? "";
        var name = item.Str("name") ?? "";
        var communityId = item.Str("community_identifier") ?? community;
        DateTime? updated = DateTime.TryParse(item.Str("last_updated") ?? item.Str("version_created"), out var d) ? d.ToUniversalTime() : null;
        return new ModInfo
        {
            Source = "thunderstore",
            Id = $"{ns}-{name}",
            Name = name.Replace('_', ' '),
            Author = ns,
            Version = item.Str("latest_version_number") ?? "",
            Description = item.Str("description") ?? "",
            Icon = item.Str("icon_url"),
            Url = $"{Base}/c/{communityId}/p/{ns}/{name}/",
            Downloads = item.Long("download_count"),
            Rating = item.Long("rating_count"),
            UpdatedAt = updated,
            Categories = item.Arr("categories").Select(c => c.Str("name")).OfType<string>().ToArray(),
        };
    }

    public static async Task<Page> Search(string community, Query q, string[] categorySlugs, CancellationToken ct = default)
    {
        var args = new List<string> { $"page={q.Page}", $"ordering={Ordering(q.Sort)}" };
        if (!string.IsNullOrWhiteSpace(q.Text)) args.Add("q=" + Uri.EscapeDataString(q.Text.Trim()));
        if (categorySlugs.Length > 0)
        {
            var ids = await CategoryIds(community, categorySlugs, ct);
            if (ids.Count == 0) return new Page([], 0, false, q.Page);
            args.AddRange(ids.Select(id => "included_categories=" + id));
        }
        var data = await Http.GetJson($"{Base}/api/cyberstorm/listing/{Uri.EscapeDataString(community)}/?{string.Join('&', args)}", ct, 20);
        var mods = data.Arr("results")
            .Where(i => i is not null && !i.Bool("is_nsfw") && !i.Bool("is_deprecated"))
            .Select(i => ToMod(i!, community))
            .ToList();
        return new Page(mods, data.Long("count"), data?["next"] is JsonValue, q.Page);
    }

    static async Task<List<string>> CategoryIds(string community, string[] slugs, CancellationToken ct)
    {
        if (!CategoryMaps.TryGetValue(community, out var map))
        {
            var data = await Http.GetJson($"{Base}/api/cyberstorm/community/{Uri.EscapeDataString(community)}/filters/", ct, 20);
            map = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            foreach (var c in data.Arr("package_categories"))
            {
                var slug = c.Str("slug");
                var id = c?["id"]?.ToString();
                if (slug is not null && id is not null) map[slug] = id;
            }
            CategoryMaps[community] = map;
        }
        return slugs.Select(s => map.GetValueOrDefault(s)).OfType<string>().ToList();
    }

    public static async Task<ModInfo?> Get(string community, string id, CancellationToken ct = default)
    {
        if (Split(id) is not var (ns, name)) return null;
        var key = $"{community}/{id}";
        if (Details.TryGetValue(key, out var hit) && DateTime.UtcNow - hit.At < TimeSpan.FromMinutes(15)) return hit.Mod;

        await Parallel.WaitAsync(ct);
        JsonNode? data;
        try
        {
            data = await Http.GetJson($"{Base}/api/cyberstorm/listing/{Uri.EscapeDataString(community)}/{Uri.EscapeDataString(ns)}/{Uri.EscapeDataString(name)}/", ct, 20);
        }
        catch (HttpRequestException e) when (e.Message.StartsWith("404")) { return null; }
        finally { Parallel.Release(); }
        if (data is null) return null;

        var mod = ToMod(data, community);
        mod.DownloadUrl = data.Str("download_url") ?? $"{Base}/package/download/{ns}/{name}/{mod.Version}/";
        mod.Dependencies = data.Arr("dependencies")
            .Where(d => d.Str("namespace") is not null && d.Str("name") is not null && !d.Bool("is_removed"))
            .Select(d => $"{d.Str("namespace")}-{d.Str("name")}")
            .ToList();
        Details[key] = (DateTime.UtcNow, mod);
        return mod;
    }

    /// <summary>Мод и все его зависимости в порядке установки (сначала зависимости).</summary>
    public static async Task<(List<ModInfo> Order, List<string> Missing)> Resolve(string community, string rootId, CancellationToken ct = default)
    {
        var order = new List<ModInfo>();
        var missing = new List<string>();
        var tasks = new Dictionary<string, Task>(StringComparer.OrdinalIgnoreCase);
        var gate = new object();

        Task Visit(string id, ImmutableTrail trail)
        {
            if (trail.Contains(id)) return Task.CompletedTask;
            lock (gate)
            {
                if (tasks.TryGetValue(id, out var existing)) return existing;
                var task = Run();
                tasks[id] = task;
                return task;
            }

            async Task Run()
            {
                var mod = await Get(community, id, ct).ConfigureAwait(false);
                if (mod is null) { lock (gate) missing.Add(id); return; }
                await Task.WhenAll(mod.Dependencies.Select(dep => Visit(dep, trail.With(id))));
                lock (gate) order.Add(mod);
            }
        }

        await Visit(rootId, ImmutableTrail.Empty);
        return (order, missing);
    }

    sealed class ImmutableTrail
    {
        public static readonly ImmutableTrail Empty = new(null, "");
        readonly ImmutableTrail? _parent;
        readonly string _id;
        ImmutableTrail(ImmutableTrail? parent, string id) { _parent = parent; _id = id; }
        public ImmutableTrail With(string id) => new(this, id);
        public bool Contains(string id)
        {
            for (var t = this; t is not null && t._parent is not null; t = t._parent)
                if (string.Equals(t._id, id, StringComparison.OrdinalIgnoreCase)) return true;
            return false;
        }
    }

    /// <summary>Последняя версия пакета (для загрузчика BepInExPack).</summary>
    public static async Task<(string Version, string Url)> Latest(string community, string fullName, CancellationToken ct = default)
    {
        if (Split(fullName) is not var (ns, name)) throw new ArgumentException(fullName);
        try
        {
            var one = await Http.GetJson($"{Base}/api/experimental/package/{Uri.EscapeDataString(ns)}/{Uri.EscapeDataString(name)}/", ct, 20);
            var latest = one?["latest"];
            if (latest.Str("download_url") is string url) return (latest.Str("version_number") ?? "", url);
        }
        catch when (!ct.IsCancellationRequested) { }

        var list = await Http.GetJson($"{Base}/c/{community}/api/v1/package/", ct, 45) as JsonArray;
        var pkg = list?.FirstOrDefault(p => p.Str("owner") == ns && p.Str("name") == name);
        var version = pkg.Arr("versions").FirstOrDefault();
        if (version.Str("download_url") is not string dl)
            throw new InvalidOperationException(I18n.T("err.bepinex.noPack", ("community", community), ("pkg", fullName)));
        return (version.Str("version_number") ?? "", dl);
    }
}
